engine/table/src/main/java/io/deephaven/engine/table/impl/select/analyzers/SelectAndViewAnalyzer.java

```diff
@@ -350,100 +376,208 @@ private static boolean shouldPreserve(final SelectColumn sc) {
+        return layers.size();
     }
 
-    abstract void populateModifiedColumnSetRecurse(ModifiedColumnSet mcsBuilder, Set<String> remainingDepsToSatisfy);
+    public static abstract class Layer implements LogOutputAppendable {
+
+        static final int BASE_LAYER_INDEX = 0;
+        static final int REDIRECTION_LAYER_INDEX = 1;
+
+        enum GetMode {
+            All, New, Published
+        }
+
+        /**
+         * The layerIndex is used to identify each layer uniquely within the bitsets for completion.
+         */
+        private final int layerIndex;
+
+        public Layer(int layerIndex) {
+            this.layerIndex = layerIndex;
+        }
+
+        int getLayerIndex() {
+            return layerIndex;
+        }
+
+        @Override
+        public String toString() {
+            return new LogOutputStringImpl().append(this).toString();
+        }
 
-    enum GetMode {
-        All, New, Published
+        public void startTrackingPrev() {
+            // default is that there is nothing to do
+        }
+
+        abstract Set<String> getLayerColumnNames();
+
+        abstract void populateModifiedColumnSetInReverse(
+                ModifiedColumnSet mcsBuilder,
+                Set<String> remainingDepsToSatisfy);
+
+        abstract void populateColumnSources(
+                Map<String, ColumnSource<?>> result,
+                GetMode mode);
+
+        abstract void calcDependsOn(
+                final Map<String, Set<String>> result,
+                boolean forcePublishAllSources);
+
+
+        abstract boolean allowCrossColumnParallelization();
+
+        /**
+         * Apply this update to this SelectAndViewAnalyzer.
+         *
+         * @param upstream the upstream update
+         * @param toClear rows that used to exist and no longer exist
+         * @param helper convenience class that memoizes reusable calculations for this update
+         * @param jobScheduler scheduler for parallel sub-tasks
+         * @param liveResultOwner {@link LivenessNode node} to be used to manage/unmanage results that happen to be
+         *        {@link io.deephaven.engine.liveness.LivenessReferent liveness referents}
+         * @param onCompletion called when the inner column is complete
+         */
+        public abstract CompletionHandler createUpdateHandler(
+                TableUpdate upstream,
+                RowSet toClear,
+                UpdateHelper helper,
+                JobScheduler jobScheduler,
+                @Nullable LivenessNode liveResultOwner,
+                CompletionHandler onCompletion);
+
+        /**
+         * A class that handles the completion of one select column.
+         */
+        public static abstract class CompletionHandler {
+            /**
+             * Note that the completed columns are shared among the entire operation's completion handlers.
+             */
+            private final BitSet completedColumns;
+            private final BitSet requiredColumns;
```

Comment:
Are these columns, or layers?

Completed columns is shared, mutable. Required is an immutable constant that belongs to the layer itself, and maybe shouldn't be reproduced here?

I wonder what we need handlers for?

We need to dispatch any layers whose required bitset is full satisfied by the completed bitset, in a parallelization rules-respecting way.